File: benchmark/comparison/tasks/22/solution.py
class Lit:
    def __init__(self, value):
        self.value = value

class Add:
    def __init__(self, left, right):
        self.left = left
        self.right = right

class Mul:
    def __init__(self, left, right):
        self.left = left
        self.right = right

def eval_expr(expr):
    if isinstance(expr, Lit):
        return expr.value
    elif isinstance(expr, Add):
        return eval_expr(expr.left) + eval_expr(expr.right)
    elif isinstance(expr, Mul):
        return eval_expr(expr.left) * eval_expr(expr.right)

def show_expr(expr):
    if isinstance(expr, Lit):
        return str(expr.value)
    elif isinstance(expr, Add):
        return f"{show_expr(expr.left)} + {show_expr(expr.right)}"
    elif isinstance(expr, Mul):
        return f"{show_expr(expr.left)} * {show_expr(expr.right)}"

print(eval_expr(Lit(5)))
print(eval_expr(Add(Lit(3), Lit(4))))
print(eval_expr(Mul(Lit(3), Lit(4))))
print(eval_expr(Add(Lit(2), Lit(3))))
print(show_expr(Add(Lit(3), Lit(4))))
print(show_expr(Mul(Lit(3), Lit(4))))

File: benchmark/comparison/tasks/22/solution.ts
type Expr =
  | { tag: "Lit"; value: number }
  | { tag: "Add"; left: Expr; right: Expr }
  | { tag: "Mul"; left: Expr; right: Expr };

function evaluate(expr: Expr): number {
  switch (expr.tag) {
    case "Lit":
      return expr.value;
    case "Add":
      return evaluate(expr.left) + evaluate(expr.right);
    case "Mul":
      return evaluate(expr.left) * evaluate(expr.right);
  }
}

function show(expr: Expr): string {
  switch (expr.tag) {
    case "Lit":
      return String(expr.value);
    case "Add":
      return `${show(expr.left)} + ${show(expr.right)}`;
    case "Mul":
      return `${show(expr.left)} * ${show(expr.right)}`;
  }
}

const lit3: Expr = { tag: "Lit", value: 3 };
const lit4: Expr = { tag: "Lit", value: 4 };
const lit5: Expr = { tag: "Lit", value: 5 };

console.log(evaluate(lit5));
console.log(evaluate({ tag: "Add", left: lit3, right: lit4 }));
console.log(evaluate({ tag: "Mul", left: lit3, right: lit4 }));
console.log(evaluate({ tag: "Add", left: { tag: "Lit", value: 2 }, right: lit3 }));
console.log(show({ tag: "Add", left: lit3, right: lit4 }));
console.log(show({ tag: "Mul", left: lit3, right: lit4 }));
